File: sdk/zig/magnolia/args.zig
pub const Args = struct {
    argc: c_int,
    argv: [*]?[*:0]u8,
    i: c_int = 0,

    pub fn init(argc: c_int, argv: [*]?[*:0]u8) Args {
        return .{ .argc = argc, .argv = argv, .i = 0 };
    }

    pub fn next(self: *Args) ?[*:0]const u8 {
        if (self.i >= self.argc) return null;
        const p = self.argv[@as(usize, @intCast(self.i))] orelse return null;
        self.i += 1;
        return @as([*:0]const u8, p);
    }
};

pub fn zlen(s: [*:0]const u8) usize {
    var n: usize = 0;
    while (s[n] != 0) : (n += 1) {}
    return n;
}

pub fn zslice(s: [*:0]const u8) []const u8 {
    return s[0..zlen(s)];
}

File: sdk/rust/magnolia-applet/src/args.rs
use core::ffi::CStr;

use crate::sys::c_char;

#[derive(Copy, Clone)]
pub struct Args {
    argc: usize,
    argv: *const *const c_char,
}

impl Args {
    pub unsafe fn from_raw(argc: i32, argv: *const *const c_char) -> Self {
        let argc = if argc > 0 { argc as usize } else { 0 };
        Self { argc, argv }
    }

    pub fn len(&self) -> usize {
        self.argc
    }

    pub fn is_empty(&self) -> bool {
        self.argc == 0
    }

    pub fn get(&self, index: usize) -> Option<&CStr> {
        if index >= self.argc {
            return None;
        }
        if self.argv.is_null() {
            return None;
        }
        unsafe {
            let ptr = *self.argv.add(index);
            if ptr.is_null() {
                None
            } else {
                Some(CStr::from_ptr(ptr))
            }
        }
    }

    pub fn iter(&self) -> ArgsIter<'_> {
        ArgsIter { args: self, index: 0 }
    }
}

pub struct ArgsIter<'a> {
    args: &'a Args,
    index: usize,
}

impl<'a> Iterator for ArgsIter<'a> {
    type Item = &'a CStr;

    fn next(&mut self) -> Option<Self::Item> {
        let item = self.args.get(self.index)?;
        self.index += 1;
        Some(item)
    }
}

